File: 2025-03-10/Balanced_Binary_Tree.java
public class Balanced_Binary_Tree {
    private static boolean ans;
    public static int solve(TreeNode root) {
        if(root == null) return 0;

        int l = solve(root.left);
        int r = solve(root.right);

        if(Math.abs(l-r) > 1)
            ans = false;

        return 1 + Math.max(l, r);
    }
    public static boolean isBalanced(TreeNode root) {
        ans = true;
        solve(root);
        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{1,2,2,3,3,null,null,4,4};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        boolean ans = isBalanced(root);
        System.out.println(ans);
    }
}

File: 2025-09-30/Balanced_Binary_Tree.py
from collections import deque
from typing import Optional, List

from TreeNode import TreeNode
from Tree import Tree


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> int:
        if not root:
            return True

        is_balanced = [True]
        def dfs(root, curDepth):
            if not root:
                return curDepth-1

            l = dfs(root.left, curDepth+1)
            r = dfs(root.right, curDepth+1)

            if is_balanced[0] and abs(l-r) > 1:
                is_balanced[0] = False

            return max(l, r)

        dfs(root, 1)
        return is_balanced[0]



sol = Solution()
tree = Tree()
node = tree.array_to_tree([1,2,2,3,None,None,3,4,None,None,4])
print(sol.isBalanced(node))
